### PR TITLE
modularize domain decomposition

### LINKED ISSUE #242: move domain decomposition into model construction
Currently domain decomposition is performed before model construction, and a simple partition of the cell ids across domains is passed into the model constructor.

To handle different cell types, internal to external gid mappings, and load balancing of cells with different computational weights, the model constructor will have to handle domain decomposition.

This task performs the first step: moving the domain decomposition logic that is currently implemented (a naive partition of gid across the domains) into the model constructor.

### PR DESCRIPTION
**Pull request by bcumming**
The domain decomposition, whereby cells were partitioned across domains/MPI
ranks and then grouped together, was performed in an ad-hoc manner.
This PR modularizes the domain decomposition.

* A `domain_decomposition` class performs the cell partitioning given the recipe
  and some parameters.
* This implementation has a single `domain_decomposition` class and flags are 
  provided in a `group_rules` struct
  * The `domain_decomposition` could be specialized, or the simple rules struct
    could provide some sort of policy implementation in the future when the 
    need arises
* The domain_decomposition class is initialized inside the model constructor
  and is maintained as state of the model.
* The cell model constructor interface has been simplified to a single constructor
  that takes a recipe and a reference to a `domain_decomposition`.
  * In the future we might pass in only a moveable subset of the 
    `domain_decomposition` information required to build the model.                                      

fixes #242.

### CODE REVIEW

**Comment by bcumming** [comment]
I am looking for some feedback here.
It has unit tests for using without MPI, however I have only got placeholders for the MPI tests.
All these tests will become much richer when we have multiple cell types and gid renumbering.

**Comment by halfflat** [comment]
Currently debating design with @bcumming IRL ...

**Comment by halfflat** [comment]
In terms of the design, I am counter-proposing the following:
* The `domain_decomposition` object is instantiated at the application level.
* `model` gets passed only the `recipe` object and a `group_allocation` object which comprises
  * A cell gid partition (representing map from local group index to cell gid interval).
  * A map from group index to (suggested) hardware back-end.

The `model` code then remains very nearly unchanged; it gets the `cell_kind` for any particular group from the recipe directly.

Why keep them separate? I believe it is a cleaner, tighter design, for the following reasons:
* The domain decomposition can be queried, or potentially tweaked, at the application level independently of model creation.
* Querying the decomposition gives the option of the application-level code pre-allocating or configuring structures, i/o etc. just for the cells that are process-local, again without needing to first instantiate the model.
* Keeping decomposition information queries outside of the `model` object reduces the size of the `model` interface. 
* The model takes a more concrete data structure for the decomposition, instead of wrapping a more opaque object whose behaviour needs to be understood in conjunction with the model initialization code.
* If a more complicated domain decomposition process requires global communication, then that communication point would be explicit in the application code rather than implicit in the model constructor.
* It makes it easy to instantiate multiple models over arbitrary and explicit ranges of cells, simplifying testing code.

File: miniapp/miniapp.cpp [hunk 1]
```diff
@@ -111,8 +104,7 @@ int main(int argc, char** argv) {
         m.set_binning_policy(binning_policy, options.bin_dt);
 
         // Inject some artificial spikes, 1 per 20 neurons.
-        cell_gid_type first_spike_cell = 20*((cell_range.first+19)/20);
-        for (auto c=first_spike_cell; c<cell_range.second; c+=20) {
+        for (cell_gid_type c=0; c<recipe->num_cells(); c+=20) {
```
Comment (bcumming) [inline miniapp/miniapp.cpp]: Yes. This interface will vanish, because it was just a hack to run benchmarks.

File: src/domain_decomposition.hpp [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+#include <util/optional.hpp>
+#include <util/partition.hpp>
+#include <util/transform.hpp>
+
+namespace nest {
+namespace mc {
+
+// Meta data used to guide the domain_decomposition in distributing
+// and grouping cells.
+struct group_rules {
```
Comment (bcumming) [inline src/domain_decomposition.hpp]: This contains information required to perform the domain decomposition. It doesn't require cell type information: such information is needed in the output of the domain decompositon, where it is used to build the concrete cell groups.

File: src/domain_decomposition.hpp [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+        auto num_domains = communication::global_policy::size();
+        auto domain_id = communication::global_policy::id();
+
+        // partition the cells globally across the domains
+        num_global_cells_ = rec.num_cells();
+        first_cell_ = (gid_type)(num_global_cells_*(domain_id/(double)num_domains));
+        last_cell_ = (gid_type)(num_global_cells_*((domain_id+1)/(double)num_domains));
+
+        // partition the local cells into cell groups
+        if (num_local_cells()>0) {
```
Comment (bcumming) [inline src/domain_decomposition.hpp]: I have added a detailed comment that describes the algorithm.

File: src/domain_decomposition.hpp [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+            gid_type group_size = 1;
+            group_starts_.push_back(first_cell_);
+            auto group_kind = rec.get_cell_kind(first_cell_);
+            group_kinds_.push_back(group_kind);
+            gid_type gid = first_cell_+1;
```
Comment (bcumming) [inline src/domain_decomposition.hpp]: Is +1 really a magic number in this context? It is simply the next index.
I have added a comment that describes the algorithm in more detail, which hopefully makes these steps clearer.

File: src/domain_decomposition.hpp [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+            while (gid<last_cell_) {
+                auto kind = rec.get_cell_kind(gid);
+                if (kind!=group_kind || group_size>=rules.target_group_size) {
```
Comment (bcumming) [inline src/domain_decomposition.hpp]: see new comment documentation.

File: src/model.hpp [hunk 6]
```diff
@@ -220,16 +213,18 @@ class model {
 
     // only thread safe if called outside the run() method
     void add_artificial_spike(cell_member_type source, time_type tspike) {
-        current_spikes().get().push_back({source, tspike});
+        if (domain_.is_local_gid(source.gid)) {
```
Comment (w-klijn) [inline src/model.hpp]: Should we add a TODO to remove this function when we have 'real' artificial cells. To prevent that we have non-threadsave code laying about?

File: src/model.hpp [hunk 7]
```diff
@@ -43,30 +44,27 @@ class model {
...
-           const util::partition_range<Iter>& groups,
-           backend_policy policy):
-        cell_group_divisions_(groups.divisions().begin(), groups.divisions().end()),
-        backend_policy_(policy)
+    model(const recipe& rec, const domain_decomposition& decomp):
+        domain_(decomp)
     {
         // set up communicator based on partition
-        communicator_ = communicator_type(gid_partition());
+        communicator_ = communicator_type(domain_.gid_group_partition());
```
Comment (w-klijn) [inline src/model.hpp]: Very clean and clear implementation!

File: tests/global_communication/test_domain_decomposition.cpp [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+using namespace nest::mc;
+
+using communicator_type = communication::communicator<communication::global_policy>;
+
+static bool is_dry_run() {
+    return communication::global_policy::kind() ==
+        communication::global_policy_kind::dryrun;
+}
+
+TEST(domain_decomp, basic) {
```
Comment (w-klijn) [inline tests/global_communication/test_domain_decomposition.cpp]: Could you give a hint what type of tests you would expect here?
As a source code comment

File: src/domain_decomposition.hpp (hunk 2 above)
Comment (w-klijn) [inline src/domain_decomposition.hpp]: I could imagine that this rules object in the future will be expanded with function like:
cell_size get_size_for_type(cell_kind):

File: src/domain_decomposition.hpp [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+
+        auto num_domains = communication::global_policy::size();
+        auto domain_id = communication::global_policy::id();
+
+        // Partition the cells globally across the domains.
+        num_global_cells_ = rec.num_cells();
+        cell_begin_ = (cell_gid_type)(num_global_cells_*(domain_id/(double)num_domains));
+        cell_end_ = (cell_gid_type)(num_global_cells_*((domain_id+1)/(double)num_domains));
+
+        // Partition the local cells into cell groups that satisfy three
```
Comment (w-klijn) [inline src/domain_decomposition.hpp]: I completely understand this now: thanx!

It even hints at possible refactorings

File: src/domain_decomposition.hpp [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+    group_range_type get_group(cell_size_type i) const {
+        return {group_starts_[i], group_starts_[i+1], group_kinds_[i]};
+    }
+
+    /// Tests whether a gid is on the local domain.
+    bool is_local_gid(cell_gid_type i) const {
+        return i>=cell_begin_ && i<cell_end_;
+    }
+
+    gid_partition_type gid_group_partition() const {
```
Comment (w-klijn) [inline src/domain_decomposition.hpp]: This function is not documented but it is maybe the most important one!

File: tests/unit/test_domain_decomposition.cpp [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+        EXPECT_TRUE(decomp.is_local_gid(i));
+    }
+    EXPECT_FALSE(decomp.is_local_gid(num_cells));
+}
+
+TEST(domain_decomposition, multi_cell_groups)
+{
+    unsigned num_cells = 10;
+
+    // test group sizes from 1 up to 1 more than the total number of cells
```
Comment (w-klijn) [inline tests/unit/test_domain_decomposition.cpp]: Would it be smart to test the beginning also? So add 2 and the first and last should not be in the group.
That way all edge cases are covered

File: src/model.hpp (hunk 6 above)
Comment (halfflat) [inline src/model.hpp]: I think there is no need at this point — it does what it says on the tin. It's likely this will go away when we have artificial spike sources in the model, but maybe it'd be useful to keep for tests. Let's not anticipate that yet :)
